ergo-lib/src/wallet/multi_sig.rs

```diff
@@ -0,0 +1,906 @@
+            .cloned()
+            .ok_or_else(|| {
+                TxSigningError::ContextError("self_index is out of bounds".to_string())
+            })?;
+        let outputs = tx
+            .output_candidates
+            .iter()
+            .enumerate()
+            .map(|(idx, b)| ErgoBox::from_box_candidate(b, tx.id(), idx as u16))
+            .collect::<Result<Vec<ErgoBox>, SigmaSerializationError>>()?;
+
+        let outputs_ir = outputs.into_iter().map(Rc::new).collect();
+        let data_inputs_ir = if let Some(data_inputs) = tx.data_inputs.as_ref() {
+            Some(data_inputs.clone().enumerated().try_mapped(|(idx, di)| {
+                data_boxes
+                    .as_ref()
+                    .ok_or(TxSigningError::DataInputBoxNotFound(idx))?
+                    .iter()
+                    .find(|b| di.box_id == b.box_id())
+                    .map(|b| Rc::new(b.clone()))
+                    .ok_or(TxSigningError::DataInputBoxNotFound(idx))
+            })?)
+        } else {
+            None
+        };
+        let inputs_ir = tx
+            .inputs_ids()
+            .clone()
+            .enumerated()
+            .try_mapped(|(idx, u)| {
+                boxes_to_spend
+                    .iter()
+                    .find(|b| u == b.box_id())
+                    .map(|b| Rc::new(b.clone()))
+                    .ok_or(TxSigningError::InputBoxNotFound(idx))
+            })?;
+        let extension = tx
+            .inputs
+            .get(i)
+            .ok_or_else(|| {
+                TxSigningError::ContextError(
+                    "self_index not found in spending transaction inputs".to_string(),
+                )
+            })?
+            .spending_proof
+            .extension
+            .clone();
+        let ctx = Rc::new(Context {
+            height,
+            self_box: Rc::new(self_box),
+            outputs: outputs_ir,
+            data_inputs: data_inputs_ir,
+            inputs: inputs_ir,
+            pre_header: state_context.pre_header.clone(),
+            headers: state_context.headers.clone(),
+            extension,
+        });
+        let tree = input_box.ergo_tree.clone();
+        let proof: ProofBytes = tx.inputs.get(i).unwrap().clone().spending_proof.proof;
+        let proof: Vec<u8> = Vec::from(proof);
+        let exp = tree
+            .proposition()
+            .map_err(ProverError::ErgoTreeError)
+            .map_err(|e| TxSigningError::ProverError(e, i))?;
+        let reduction_result = reduce_to_crypto(&exp, &Env::empty(), ctx)
+            .map_err(ProverError::EvalError)
+            .map_err(|e| TxSigningError::ProverError(e, i))?;
+        let sigma_tree = reduction_result.sigma_prop;
+        let bag = bag_for_multi_sig(
+            sigma_tree,
+            real_secrets_to_extract.as_slice(),
+            simulated_secrets_to_extract.as_slice(),
+            proof.as_slice(),
+        );
```

Comment:
Let's properly propagate the error up with
```suggestion
        )?;
```
and remove `match` below which silence the error.

Reply:
fixed!